src/platform/intel/ace/platform.c

```diff
@@ -73,6 +77,59 @@ int platform_boot_complete(uint32_t boot_message)
 	return ipc_platform_send_msg(&msg);
 }
 
+/* address where zephyr PM will save memory during D3 transition */
+extern void *global_imr_ram_storage;
+
+/* notifier called before every power state transition */
+static void notify_pm_state_entry(enum pm_state state)
+{
+	size_t storage_buffer_size;
+
+	if (state == PM_STATE_SOFT_OFF) {
+		/* allocate IMR global_imr_ram_storage */
+		const struct device *tlb_dev = DEVICE_DT_GET(DT_NODELABEL(tlb));
+
+		__ASSERT_NO_MSG(tlb_dev);
+		const struct intel_adsp_tlb_api *tlb_api =
+				(struct intel_adsp_tlb_api *)tlb_dev->api;
+
+		/* get HPSRAM storage buffer size */
+		storage_buffer_size = tlb_api->get_storage_size();
+
+		/* add space for LPSRAM */
+		storage_buffer_size += LP_SRAM_SIZE;
+
+		/* allocate IMR buffer and store it in the global pointer */
+		global_imr_ram_storage = rmalloc(SOF_MEM_ZONE_SYS,
+						 0,
+						 SOF_MEM_CAPS_L3,
+						 storage_buffer_size);
```

Comment:
Should be done only for the primary core.

Reply:
And it is.
Callbacks `notify_pm_state_entry` and `notify_pm_state_exit` are registered in the body of the:
```
/* Runs on the primary core only */
int platform_init(struct sof *sof)
```
function. I'll write some docs.

Reply:
is this ok now?

Reply:
I don't see any changes. If you mean your comment then no.

This comment in the code is about the _**platform_init**_ function, and it's true. But the notifications you register are global and will be called from every core.

Reply:
Sorry, I added comments and forgot to push. Now the docs should be visible.

Reply:
This code will be called on other cores than primary. You need to add _**if (cpu_is_primary(arch_proc_id()))**_

Reply:
is this ok now?